src/shims/panic.rs

```diff
@@ -119,19 +119,54 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
     fn handle_stack_pop(
         &mut self,
-        mut extra: FrameData<'tcx>,
+        mut frame: Frame<'mir, 'tcx, Tag, FrameData<'tcx>>,
         unwinding: bool,
     ) -> InterpResult<'tcx, StackPopJump> {
         let this = self.eval_context_mut();
 
-        trace!("handle_stack_pop(extra = {:?}, unwinding = {})", extra, unwinding);
+        trace!("handle_stack_pop(frame.extra = {:?}, unwinding = {})", frame.extra, unwinding);
+
+        if unwinding {
+            for &def_id in &[frame.instance.def_id(), frame.body.source.def_id()] {
```

Comment:
Why do you (need to) check both of them?

Reply:
These two tests will fail if I remove either, but I think they should not:
https://github.com/rust-lang/miri/blob/2d00a545c1ed7f658d0dca22eee9be96764c600b/tests/compile-fail/non_extern_foreign_call_bad_unwind2.rs
https://github.com/rust-lang/miri/blob/2d00a545c1ed7f658d0dca22eee9be96764c600b/tests/compile-fail/non_extern_foreign_call_bad_unwind3.rs

By the way, I think I can write `for def_id in [..., ...]` without `&` with a newer `rustc`, but `rust-version` in my branch is too old now.

Reply:
> These two tests will fail if I remove either, but I think they should not:

I don't think this is the right fix though. What you want to check, I guess, is caller and callee ABI -- but what you are checking here is callee ABI and "import ABI", and "import ABI" really shouldn't matter. It's just that in your examples, import ABI and caller ABI are the same.
But if I import things under one ABI, then transmute the fn ptr around and call it with some other ABI -- then that other ABI is what should be checked.

Reply:
Good point. I think for the caller, only *attributes* (`#[rustc_allocator_nounwind]` and `#[unwind(allowed)]`) is needed to be checked here since the caller and callee ABI must match. But is it possible to get that information? It does not seem like stored in https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/interpret/struct.Frame.html, and the information about the original function pointer seems discarded in https://github.com/rust-lang/rust/blob/50f2bf6a5751751ea27a8fd5577d5bdd37236669/compiler/rustc_mir/src/interpret/terminator.rs#L59-L64.

Reply:
> I think for the caller, only attributes (#[rustc_allocator_nounwind] and #[unwind(allowed)]) is needed to be checked here since the caller and callee ABI must match.

Sorry for the sloppy terminology; I considered those to be somehow part of the ABI.

> But is it possible to get that information?

Not currently, but we can arrange for that I think. I think `StackPopCleanup` is the right place for this: it already supports setting `ret` to `None` to indicate "the caller does not expect this function to return"; we should similarly be able to express "the caller does not expect this function to unwind". Note that `unwind` here already is an `Option`, but currently `None` means "the caller does not have to do any cleanup, just drop that stack frame and go on".